lineapy/plugins/pipeline_writers.py

```diff
@@ -302,48 +342,172 @@ def task_iris_pred():
+                    ),
+                    typing_blocks=user_input_var_typing_block,
+                    loading_blocks=input_var_loading_block,
+                    call_block=function_call_block,
+                    dumping_blocks=return_var_saving_block,
+                    indentation_block=" " * indentation,
+                )
+                task_def: TaskDefinition = {
+                    "definition": function_definition,
+                    "user_input_variables": artifact_user_input_variables,
+                }
+                task_definitions[nc.safename] = task_def
+
+        return task_definitions
+
 
-def get_task_definition(
-    nc: NodeCollection, pipeline_name: str, indentation=4
+def get_session_task_definition(
+    sa: SessionArtifacts,
+    pipeline_name: str,
+    indentation=4,
 ) -> str:
     """
-    Add deserialization of input variables and serialization of output
-    variables logic of the call_block and wrap them into a new function
-    definition.
+    Add serialization of output artifacts logic of the session function
+    call_block and wrap them into a new function definition.
     """
-    input_var_loading_block = [
-        f"{var} = pickle.load(open('/tmp/{pipeline_name}/variable_{var}.pickle','rb'))"
-        for var in sorted(list(nc.input_variables))
+    session_input_parameters_spec = sa.get_session_input_parameters_spec()
+    session_input_variables = list(session_input_parameters_spec.keys())
+    user_input_var_typing_block = [
+        f"{var} = {session_input_parameters_spec[var].value_type}({var})"
+        for var in session_input_variables
     ]
-    function_call_block = nc.get_function_call_block(
-        indentation=0, source_module=f"{pipeline_name}_module"
-    )
-    return_var_saving_block = [
-        f"pickle.dump({var},open('/tmp/{pipeline_name}/variable_{var}.pickle','wb'))"
-        for var in nc.return_variables
+
+    input_var_loading_block: List[str] = []
+    function_call_block = f"artifacts = {pipeline_name}_module.{sa.get_session_function_callblock()}"
+    return_artifacts_saving_block = [
+        f"pickle.dump(artifacts['{nc.name}'],open('/tmp/{pipeline_name}/artifact_{nc.safename}.pickle','wb'))"
+        for nc in sa.artifact_nodecollections
+        if nc.collection_type == NodeCollectionType.ARTIFACT
     ]
 
     TASK_FUNCTION_TEMPLATE = load_plugin_template("task_function.jinja")
     return TASK_FUNCTION_TEMPLATE.render(
-        artifact_name=nc.safename,
+        function_name=sa.get_session_function_name(),
+        user_input_variables=", ".join(session_input_variables),
+        typing_blocks=user_input_var_typing_block,
         loading_blocks=input_var_loading_block,
         call_block=function_call_block,
-        dumping_blocks=return_var_saving_block,
+        dumping_blocks=return_artifacts_saving_block,
         indentation_block=" " * indentation,
     )
 
 
+class AirflowCodeGenerator:
```

Comment:
I am wondering if we can actually absorb this into existing `AirflowPipelineWriter` class. Since `AirflowPipelineWriter` already has `self.artifact_collection` attribute, all methods of `AirflowCodeGenerator` can simply be put under `AirflowPipelineWriter`.

Is there something I am missing in this thought @mingjerli?